tests/test_server.py

```diff
@@ -58,9 +57,9 @@ def test_api_empty_db(client, url, exp_res):
 @pytest.mark.parametrize(
     'url, exp_res, status_code, method', [
         ['/api/tags/1', {'message': 'This resource does not exist.'}, 404, 'get'],
-        ['/api/tags/1', response_template['failure'], 400, 'put'],
+        ['/api/tags/1', None, 400, 'put'],
         ['/api/bookmarks/1', response_template['failure'], 400, 'get'],
-        ['/api/bookmarks/1', response_template['failure'], 400, 'put'],
+        ['/api/bookmarks/1', None, 400, 'put'],
```

Comment:
I did notice too. I can investigate more.
As a comparison, if I  use curl without providing any data along with the PUT request,   I receive this response (html, not json):
```
curl -X PUT 'http://127.0.0.1:5000/api/bookmarks/1'

<!doctype html>
<html lang=en>
<title>400 Bad Request</title>
<h1>Bad Request</h1>
<p>Did not attempt to load JSON data because the request Content-Type was not &#39;application/json&#39;.</p>
```

Reply:
> Come to think of it… why is it that in these cases alone, the response is None? There's a bunch of other HTTP 400 in the test cases, but all of those return data.

All the other cases provide the json parameter to `client.put. So we are not really checking for an invalid_id in the `post` and `put` cases.

Reply:
The exception is thrown when calling `request.get_json()` while retrieving the form:
```
<class 'werkzeug.exceptions.BadRequest'>
Exception 400 Bad Request: The browser (or proxy) sent a request that this server could not understand.
 ```
I can catch it and return `response.bad()` instead.

In the `'/api/bookmarks/1/2', response_template['failure'], 400, 'put']` case the `reponse.bad()` was returned when checking for an invalid range, before trying to get the json.

Reply:
…Oddly enough this seems to be passing on py37 (but not py38, py39 nor py310).